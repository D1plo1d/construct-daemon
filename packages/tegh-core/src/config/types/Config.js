import { Record, List } from 'immutable'
import uuid from 'uuid/v4'

import HostConfig from './HostConfig'
import PrinterConfig from './PrinterConfig'
import MaterialConfig from './MaterialConfig'
import developmentConfig from '../../../../../development.config'

export const ConfigRecordFactory = Record({
  id: null,
  modelVersion: 0,
  host: null,
  printer: null,
  materials: List(),
})

const Config = ({
  id = uuid(),
  modelVersion = 0,
  host = {},
  printer = {},
  materials = [],
  ...props
} = {}) => (
  ConfigRecordFactory({
    id,
    modelVersion,
    host: HostConfig(host),
    printer: PrinterConfig(printer),
    materials: List(materials).map(material => MaterialConfig(material)),
    ...props,
  })
)

export const MockConfig = ({
  host = {},
  printer = {},
  ...props
} = {}) => (
  Config({
    ...developmentConfig,
    host: {
      ...developmentConfig.host,
      ...host,
    },
    printer: {
      ...developmentConfig.printer,
      ...printer,
    },
    ...props,
  })
)

export default Config
