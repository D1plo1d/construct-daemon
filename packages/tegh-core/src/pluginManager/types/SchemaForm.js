import { Record, Map } from 'immutable'

const SchemaForm = Record({
  id: null,
  schema: null,
  form: null,
  configPaths: Map({}),
})

export const NullSchemaForm = SchemaForm({
  id: 'NULL',
  schema: {},
  form: [],
})

export default SchemaForm
