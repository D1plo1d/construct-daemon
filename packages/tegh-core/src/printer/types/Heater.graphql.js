import tql from 'typiql'
import snl from 'strip-newlines'
import {
  GraphQLObjectType,
} from 'graphql'

import getHeaterConfigs from '../../config/selectors/getHeaterConfigs'

const HeaterType = new GraphQLObjectType({
  name: 'Heater',
  description: 'A heated bed or extruder',
  fields: () => ({
    id: {
      type: tql`ID!`,
      resolve: source => `${source.id}DynamicData`,
    },
    name: {
      type: tql`String!`,
      resolve: (source, args, { store }) => {
        const state = store.getState()
        return getHeaterConfigs(state.config).get(source.id).name
      },
    },
    type: {
      type: tql`String!`,
      resolve: (source, args, { store }) => {
        const state = store.getState()
        return getHeaterConfigs(state.config).get(source.id).type
      },
    },
    targetTemperature: {
      type: tql`Float`,
      description: snl`
        The target temperature in °C for this heater. The heater will
        attempt to make the current_temp equal to this temperature.
      `,
    },
    currentTemperature: {
      type: tql`Float!`,
      description: snl`
        The current temperature in °C recorded by the heater’s thermocouple
        or thermister.
      `,
    },
    blocking: {
      type: tql`Boolean!`,
      description: snl`
        True if the printer is waiting on this heater to reach it’s
        targetTemp and preventing any more gcodes from executing until it does.
      `,
    },
  }),
})

export default HeaterType
