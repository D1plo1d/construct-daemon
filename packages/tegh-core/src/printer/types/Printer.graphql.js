import { GraphQLObjectType } from 'graphql'
import tql from 'typiql'
import snl from 'strip-newlines'

import isIdle from '../../spool/selectors/isIdle'
import getDriverState from '../selectors/getDriverState'

import PrinterStatusEnum from './PrinterStatusEnum.graphql'
import HeaterType from './Heater.graphql'
import FanType from './Fan.graphql'
import LogEntryType from '../../log/types/LogEntry.graphql'
import MacroDefinitionType from '../../config/types/MacroDefinition.graphql'
import PrinterErrorType from './PrinterError.graphql'

const PrinterGraphQL = new GraphQLObjectType({
  name: 'Printer',
  fields: () => ({
    id: {
      type: tql`ID!`,
      resolve(source) {
        return source.config.id
      },
    },
    name: {
      type: tql`String!`,
      resolve(source) {
        return source.config.name
      },
    },
    heaters: {
      type: tql`[${HeaterType}!]!`,
      resolve(source) {
        return getDriverState(source).peripherals.heaters.values()
      },
    },
    targetTemperaturesCountdown: {
      type: tql`Float`,
      description: snl`
        The estimated number of seconds until the heater(s) reach their
        targetTemperature.
      `,
      resolve(source) {
        return getDriverState(source).peripherals.targetTemperaturesCountdown
      },
    },
    activeExtruderID: {
      type: tql`String`,
      description: snl`
        The active extruder ID
      `,
      resolve(source) {
        return getDriverState(source).peripherals.activeExtruderID
      },
    },
    fans: {
      type: tql`[${FanType}!]!`,
      resolve(source) {
        return getDriverState(source).peripherals.fans.values()
      },
    },
    status: {
      type: tql`${PrinterStatusEnum}!`,
      resolve: (source) => {
        if (!isIdle(source.spool)) return 'PRINTING'
        const { status } = source.status
        return status.substring(status.lastIndexOf('/') + 1)
      },
    },
    error: {
      type: tql`${PrinterErrorType}`,
      resolve: source => source.status.error,
    },
    macroDefinitions: {
      type: tql`[${MacroDefinitionType}!]!`,
      resolve: source => source.macros.keys(),
    },
    logEntries: {
      type: tql`[${LogEntryType}!]`,
      args: {
        level: {
          type: tql`String`,
        },
        source: {
          type: tql`String`,
        },
      },
      resolve(source, args) {
        let entries = source.log.get('entries')
        if (args.level != null) {
          entries = entries.filter(log => log.level === args.level)
        }
        if (args.source != null) {
          entries = entries.filter(log => log.source === args.source)
        }
        return entries.toArray()
      },
    },
  }),
})

export default PrinterGraphQL
