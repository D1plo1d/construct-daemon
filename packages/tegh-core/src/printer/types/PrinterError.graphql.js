import {
  GraphQLObjectType,
} from 'graphql'
import tql from 'typiql'
import snl from 'strip-newlines'

const PrinterErrorType = new GraphQLObjectType({
  name: 'PrinterError',
  fields: () => ({
    code: {
      type: tql`String!`,
      description: snl`
        A machine-readable code indicating the type of error
      `,
    },
    message: {
      type: tql`String!`,
      description: snl`
        A human-readable description of the error
      `,
    },
  }),
})

export default PrinterErrorType
