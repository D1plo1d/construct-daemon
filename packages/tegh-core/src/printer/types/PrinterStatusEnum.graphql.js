import snl from 'strip-newlines'
import {
  GraphQLEnumType,
} from 'graphql'

const PrinterStatusEnum = new GraphQLEnumType({
  name: 'PrinterStatusEnum',
  values: {
    CONNECTING: {
      description: snl`
        The printer is being initialized.
        Attempting to spool anything will result in an error.
      `,
    },
    READY: {
      description: snl`
        The printer is connected and able to spool tasks/jobs.
      `,
    },
    PRINTING: {
      description: snl`
        The printer is printing a job.
        Attempting to spool anything will result in an error.
      `,
    },
    DISCONNECTED: {
      description: snl`
        The printer is disconnected or turned off.
        Attempting to spool anything will result in an error.
      `,
    },
    ERRORED: {
      description: snl`
        The printer is being initialized.
        Attempting to spool anything except for an
        emergency macro (ie. reset) will result in an error. Spool the \`reset\`
        macro to reset the error and change the status to \`CONNECTING\`.
      `,
    },
    ESTOPPED: {
      description: snl`
        The printer is estopped. Attempting to spool anything except for an
        emergency macro (ie. reset) will result in an error. Spool the \`reset\`
        macro to reset the estop and change the status to \`CONNECTING\`.
      `,
    },
  },
})

export default PrinterStatusEnum
