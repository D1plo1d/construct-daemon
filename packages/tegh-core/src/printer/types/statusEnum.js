export const ERRORED = 'tegh/printer/status/ERRORED'
export const ESTOPPED = 'tegh/printer/status/ESTOPPED'
export const DISCONNECTED = 'tegh/printer/status/DISCONNECTED'
export const CONNECTING = 'tegh/printer/status/CONNECTING'
export const READY = 'tegh/printer/status/READY'
