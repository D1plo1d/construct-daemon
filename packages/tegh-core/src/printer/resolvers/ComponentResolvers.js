import getComponentsState from '../selectors/getComponentsState'
import { FAN, TOOLHEAD } from '../../config/types/components/ComponentTypeEnum'
import { NullSchemaForm } from '../../pluginManager/types/SchemaForm'
import getAxePositions from '../selectors/getAxePositions'

const ComponentResolvers = {
  Component: {
    name: source => (
      source.model.get('name')
    ),
    configForm: (source, args, { store }) => {
      const { id, model, modelVersion } = source

      const state = store.getState()
      const schemaForm = state.schemaForms.getIn(
        ['components', source.type],
        NullSchemaForm,
      )

      return {
        id,
        model,
        modelVersion,
        schemaForm,
      }
    },
    axis: (source, args, { store }) => {
      const state = store.getState()

      const address = source.model.get('address')

      return {
        id: address,
        position: getAxePositions(state)[address],
      }
    },
    address: source => (
      source.model.get('address')
    ),
    heater: (source, args, { store }) => {
      const heater = source.getIn(['model', 'heater'])

      if (heater) {
        const state = store.getState()
        const address = source.model.get('address')
        const dynamicComponent = getComponentsState(state).byAddress.get(address)
        return dynamicComponent
      }

      return null
    },
    toolhead: (source) => {
      if (source.type === TOOLHEAD) {
        return source
      }

      return null
    },
    fan: (source, args, { store }) => {
      if (source.type === FAN) {
        const state = store.getState()
        const address = source.model.get('address')
        const dynamicComponent = getComponentsState(state).byAddress.get(address)
        return dynamicComponent
      }

      return null
    },
  },
}

export default ComponentResolvers
