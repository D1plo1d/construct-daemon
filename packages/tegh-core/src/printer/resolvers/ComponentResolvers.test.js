import { Map, Record } from 'immutable'
import typeDefs from 'tegh-schema'

import deterministicTestSetup from '../../util/testing/deterministicTestSetup'
import snapshotTestResolvers from '../../util/testing/snapshotTestResolvers'

import { MockConfig } from '../../config/types/Config'

import ComponentResolvers from './ComponentResolvers'
import SchemaForm from '../../pluginManager/types/SchemaForm'

describe('ComponentResolvers', () => {
  it('presents a consistent API', () => {
    deterministicTestSetup()

    const state = Record({
      driver: 'tegh-driver-serial-gcode',
      config: MockConfig(),
      schemaForms: Record({
        components: Map({
          TOOLHEAD: SchemaForm({}),
        }),
      })(),
      'tegh-driver-serial-gcode': {
        components: Record({
          byAddress: Map({
            e1: Record({
              id: 'heater_1',
              address: 'e1',
              currentTemperature: 32,
              targetTemperature: 40,
              blocking: true,
            })(),
          }),
        })(),
      },
    })()

    snapshotTestResolvers({
      typeDefs,
      typeName: 'Component',
      resolvers: ComponentResolvers,
      rootValue: state.config.printer.components.find(c => c.id === 'heater_1'),
      contextValue: {
        store: {
          getState: () => state,
        },
      },
    })
  })
})
