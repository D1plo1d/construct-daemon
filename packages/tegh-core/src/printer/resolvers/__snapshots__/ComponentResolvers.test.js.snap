// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ComponentResolvers presents a consistent API 1`] = `
Object {
  "_testComponent": Object {
    "address": "e0",
    "configForm": Object {
      "id": "heater_1",
    },
    "fan": null,
    "heater": null,
    "id": "heater_1",
    "name": "Extruder 1",
    "type": "TOOLHEAD",
  },
}
`;
